### DISCUSSION TITLE
Any method to run mongoose in a single thread?

### DISCUSSION THREAD

**Discussion by zeroxer** [cesanta/mongoose#2177] [Ask questions, get answers] [answered]
I want to start a http server in a seprate thread in my app, not the main thread. Can mongoose works fine in this way? Any example code will be helpful.

I have tried this, but it doesn't work. And I cannot open the http://127.0.0.1:8000.
I have tried use packed file in c(main.c) according to the documentation of mongoose and it works fine.
- main.cpp
```cpp

#include "mongoose.h"

const char *s_listening_url = "http://0.0.0.0:8000";

// HTTP request handler function
void fn(struct mg_connection *c, int ev, void *ev_data, void *fn_data)
{
    if (ev == MG_EV_HTTP_MSG)
    {
        struct mg_http_serve_opts opts;
        opts.root_dir = "/dist";
        opts.fs = &mg_fs_packed;
        mg_http_serve_dir(
            static_cast<struct mg_connection *>(c),
            static_cast<struct mg_http_message *>(ev_data),
            static_cast<struct mg_http_serve_opts *>(&opts));
    }
    // (void)fn_data;
}

void http_main(void)
{
    struct mg_mgr mgr;
    mg_log_set(MG_LL_INFO);
    mg_mgr_init(&mgr);
    mg_http_listen(&mgr, s_listening_url, fn, NULL);
    while (true)
        mg_mgr_poll(&mgr, 500);
    mg_mgr_free(&mgr);
}

int main() {
    std::thread http_thread(http_main);
    http_thread.detach();
   // main logic of other functions
   // ......
    return 0;
}
```

**Comment by scaprile** [comment, marked as answer]
Please read the docs and try before opening an _issue_. If you want to ask a _question_ there is the discussion area.

From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above), examples, and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do multi-threading.

1. From the examples (in this very repo): the new way, use a queue
https://github.com/cesanta/mongoose/tree/master/examples/multi-threaded

2. From the tutorials: the old way, the socket pipe
https://mongoose.ws/documentation/tutorials/multi-threaded/
This will probably be deprecated in upcoming releases, find the example in 7.9

3. Use your OS/language inter-thread/process messaging mechanisms between your event handlers and your worker threads.

**Reply by scaprile** [reply]
BTW, "all calls in the same thread" means that, if you want to run Mongoose in a single thread then "run Mongoose in a single thread": check out the FreeRTOS (Keil), Zephyr and RTX (Keil) examples for the STM32, those **start the thread first, initialize everything, and then loop**; this way all Mongoose calls are in the very same thread.
Or, spawn other tasks and keep Mongoose main loop in the main thread.

**Comment by zeroxer** [comment]
Thanks
